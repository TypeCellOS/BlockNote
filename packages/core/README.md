<p align="center">
  <a href="https://www.blocknotejs.org">
    <img alt="TypeCell" src="https://github.com/TypeCellOS/BlockNote/raw/main/docs/public/img/logos/banner.svg?raw=true" width="300" />
  </a>
</p>

<p align="center">
Welcome to BlockNote! The open source Block-Based
React rich text editor. Easily add a modern text editing experience to your app.
</p>

<p align="center">
  <a href="https://www.blocknotejs.org">
    Homepage
  </a> - <a href="https://www.blocknotejs.org/docs">
    Documentation
  </a> - <a href="https://www.blocknotejs.org/docs/getting-started">
    Quickstart
  </a>- <a href="https://www.blocknotejs.org/examples">
    Examples
  </a>
</p>

# Live demo

See our homepage @ [https://www.blocknotejs.org](https://www.blocknotejs.org/) or browse the [examples](https://www.blocknotejs.org/examples).

# Example code (React)

[![npm version](https://badge.fury.io/js/%40blocknote%2Freact.svg)](https://badge.fury.io/js/%40blocknote%2Freact)

```typescript
import { useCreateBlockNote } from "@blocknote/react";
import { BlockNoteView } from "@blocknote/mantine";
import "@blocknote/core/fonts/inter.css";
import "@blocknote/mantine/style.css";

function App() {
  const editor = useCreateBlockNote();

  return <BlockNoteView editor={editor} />;
}
```

`@blocknote/react` comes with a fully styled UI that makes it an instant, polished editor ready to use in your app.

# Features

BlockNote comes with a number of features and components to make it easy to embed a high-quality block-based editor in your app:

### Animations:

<img src="https://github.com/TypeCellOS/BlockNote/blob/readme/.resources/animations.gif?raw=true" width="400" />

### Helpful placeholders:

<img src="https://github.com/TypeCellOS/BlockNote/blob/readme/.resources/placeholders.gif?raw=true" width="400" />

### Drag and drop blocks:

<img src="https://github.com/TypeCellOS/BlockNote/blob/readme/.resources/dragdrop.gif?raw=true" width="400" />

### Nesting / indentation with tab and shift+tab:

<img src="https://github.com/TypeCellOS/BlockNote/blob/readme/.resources/nesting.gif?raw=true" width="400" />

### Slash (/) menu:

<img src="https://github.com/TypeCellOS/BlockNote/blob/readme/.resources/slashmenu.gif?raw=true" width="400" />

### Format menu:

<img src="https://github.com/TypeCellOS/BlockNote/blob/readme/.resources/selectmenu.gif?raw=true" width="400" />

### Real-time collaboration:

<img src="https://github.com/TypeCellOS/BlockNote/blob/readme/.resources/collaboration.gif?raw=true" width="400" />

# Feedback üôã‚Äç‚ôÇÔ∏èüôã‚Äç‚ôÄÔ∏è

We'd love to hear your thoughts and see your experiments, so [come and say hi on Discord](https://discord.gg/Qc2QTTH5dF).

# Contributing üôå

See [CONTRIBUTING.md](CONTRIBUTING.md) for more info and guidance on how to run the project (TLDR: just use `pnpm start`).

The codebase is automatically tested using Vitest and Playwright.

# License üìÉ

BlockNote is 100% Open Source Software. The majority of BlockNote is licensed under the [MPL-2.0 license](LICENSE-MPL.txt), which allows you to use BlockNote in commercial (and closed-source) applications. If you make changes to the BlockNote source files, you're expected to publish these changes so the wider community can benefit as well. [Learn more](https://fossa.com/blog/open-source-software-licenses-101-mozilla-public-license-2-0/).

The XL packages (source code in the `packages/xl-*` directories and published in NPM as `@blocknote/xl-*`) are licensed under the AGPL-3.0. If you cannot comply with this license and want to use the XL libraries, you'll need a commercial license. Refer to [our website](https://www.blocknotejs.org/pricing) for more information.

# Credits ‚ù§Ô∏è

BlockNote builds directly on two awesome projects; [Prosemirror](https://prosemirror.net/) by Marijn Haverbeke and [Tiptap](https://tiptap.dev/). Consider sponsoring those libraries when using BlockNote: [Prosemirror](https://marijnhaverbeke.nl/fund/), [Tiptap](https://github.com/sponsors/ueberdosis).

BlockNote is built as part of [TypeCell](https://www.typecell.org). TypeCell is proudly sponsored by the renowned [NLNet foundation](https://nlnet.nl/foundation/) who are on a mission to support an open internet, and protect the privacy and security of internet users. Check them out!

<a href="https://nlnet.nl"><img src="https://nlnet.nl/image/logos/NGIAssure_tag.svg" alt="NLNet" width="100"></a>

Hosting and deployments powered by Vercel:

<a href="https://vercel.com/?utm_source=TypeCell&utm_campaign=oss"><img src="https://images.ctfassets.net/e5382hct74si/78Olo8EZRdUlcDUFQvnzG7/fa4cdb6dc04c40fceac194134788a0e2/1618983297-powered-by-vercel.svg" alt="NLNet" width="150"></a>

This project is tested with BrowserStack
