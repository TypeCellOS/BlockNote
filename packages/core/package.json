{
  "name": "@blocknote/core",
  "homepage": "https://github.com/TypeCellOS/BlockNote",
  "private": false,
  "license": "MPL-2.0",
  "version": "0.9.3",
  "files": [
    "dist",
    "types",
    "src"
  ],
  "keywords": [
    "react",
    "javascript",
    "editor",
    "typescript",
    "prosemirror",
    "wysiwyg",
    "rich-text-editor",
    "notion",
    "yjs",
    "block-based",
    "tiptap"
  ],
  "description": "A \"Notion-style\" block-based extensible text editor built on top of Prosemirror and Tiptap.",
  "type": "module",
  "source": "src/index.ts",
  "types": "./types/src/index.d.ts",
  "main": "./dist/blocknote.umd.cjs",
  "module": "./dist/blocknote.js",
  "exports": {
    ".": {
      "types": "./types/src/index.d.ts",
      "import": "./dist/blocknote.js",
      "require": "./dist/blocknote.umd.cjs"
    },
    "./style.css": {
      "import": "./dist/style.css",
      "require": "./dist/style.css"
    }
  },
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "build-bundled": "tsc && vite build --config vite.config.bundled.ts && git checkout tmp-releases && rm -rf ../../release && mv ../../release-tmp ../../release",
    "preview": "vite preview",
    "lint": "eslint src --max-warnings 0",
    "test": "vitest --run",
    "test-watch": "vitest watch"
  },
  "dependencies": {
    "@emotion/cache": "^11.10.5",
    "@emotion/serialize": "^1.1.1",
    "@emotion/utils": "^1.2.0",
    "@tiptap/core": "^2.0.3",
    "@tiptap/extension-bold": "^2.0.3",
    "@tiptap/extension-code": "^2.0.3",
    "@tiptap/extension-collaboration": "^2.0.3",
    "@tiptap/extension-collaboration-cursor": "^2.0.3",
    "@tiptap/extension-dropcursor": "^2.0.3",
    "@tiptap/extension-gapcursor": "^2.0.3",
    "@tiptap/extension-hard-break": "^2.0.3",
    "@tiptap/extension-history": "^2.0.3",
    "@tiptap/extension-horizontal-rule": "^2.0.3",
    "@tiptap/extension-italic": "^2.0.3",
    "@tiptap/extension-link": "^2.0.3",
    "@tiptap/extension-paragraph": "^2.0.3",
    "@tiptap/extension-strike": "^2.0.3",
    "@tiptap/extension-text": "^2.0.3",
    "@tiptap/extension-underline": "^2.0.3",
    "@tiptap/pm": "^2.0.3",
    "hast-util-from-dom": "^4.2.0",
    "lodash": "^4.17.21",
    "prosemirror-model": "^1.18.3",
    "prosemirror-state": "^1.4.3",
    "prosemirror-transform": "^1.7.2",
    "prosemirror-view": "^1.31.4",
    "rehype-parse": "^8.0.4",
    "rehype-remark": "^9.1.2",
    "rehype-stringify": "^9.0.3",
    "remark-gfm": "^3.0.1",
    "remark-parse": "^10.0.1",
    "remark-rehype": "^10.1.0",
    "remark-stringify": "^10.0.2",
    "unified": "^10.1.2",
    "uuid": "^8.3.2",
    "y-prosemirror": "1.0.20",
    "y-protocols": "^1.0.5",
    "yjs": "^13.6.1"
  },
  "devDependencies": {
    "@types/hast": "^2.3.4",
    "@types/lodash": "^4.14.179",
    "@types/uuid": "^8.3.4",
    "eslint": "^8.10.0",
    "jsdom": "^21.1.0",
    "prettier": "^2.7.1",
    "typescript": "^5.0.4",
    "vite": "^4.4.8",
    "vite-plugin-eslint": "^1.8.1",
    "vitest": "^0.34.1"
  },
  "eslintConfig": {
    "extends": [
      "../../.eslintrc.js"
    ]
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "gitHead": "37614ab348dcc7faa830a9a88437b37197a2162d"
}
