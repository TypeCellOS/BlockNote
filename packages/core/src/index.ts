export * from "./api/exporters/html/externalHTMLExporter";
export * from "./api/exporters/html/internalHTMLSerializer";
export * from "./api/testUtil";
export * from "./blocks/ImageBlockContent/uploadToTmpFilesDotOrg_DEV_ONLY";
export * from "./blocks/defaultBlocks";
export * from "./blocks/defaultProps";
export * from "./editor/BlockNoteEditor";
export * from "./editor/BlockNoteExtensions";
export * from "./editor/selectionTypes";
export * from "./extensions-shared/BaseUiElementTypes";
export type { SuggestionItem } from "./extensions-shared/suggestion/SuggestionItem";
export * from "./extensions-shared/suggestion/SuggestionPlugin";
export * from "./extensions/FormattingToolbar/FormattingToolbarPlugin";
export * from "./extensions/HyperlinkToolbar/HyperlinkToolbarPlugin";
export * from "./extensions/ImageToolbar/ImageToolbarPlugin";
export * from "./extensions/SideMenu/SideMenuPlugin";
export * from "./extensions/SlashMenu/BaseSlashMenuItem";
export * from "./extensions/SlashMenu/SlashMenuPlugin";
export { getDefaultSlashMenuItems } from "./extensions/SlashMenu/defaultSlashMenuItems";
export * from "./extensions/TableHandles/TableHandlesPlugin";
export * from "./schema";
export * from "./util/browser";
export * from "./util/string";
// for testing from react (TODO: move):
export * from "./api/nodeConversions/nodeConversions";
export * from "./api/testUtil/partialBlockTestUtil";
export * from "./extensions/UniqueID/UniqueID";
export { UnreachableCaseError } from "./util/typescript";
