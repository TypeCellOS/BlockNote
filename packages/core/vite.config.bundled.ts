import * as path from "path";
import { defineConfig } from "vite";

// import eslintPlugin from "vite-plugin-eslint";

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [],
  build: {
    outDir: "../../release-tmp",
    minify: false,
    sourcemap: true,
    lib: {
      entry: path.resolve(__dirname, "src/index.ts"),
      name: "blocknote",
      fileName: "blocknote.bundled",
    },
    rollupOptions: {
      // external: Object.keys(pkg.dependencies),
      output: {
        // Provide global variables to use in the UMD build
        // for externalized deps
        globals: {
          // react: "React",
          // "react-dom": "ReactDOM",
        },
      },
    },
  },
});
