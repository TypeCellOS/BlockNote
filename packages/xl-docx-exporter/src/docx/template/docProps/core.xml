<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<cp:coreProperties
    xmlns:cp="http://schemas.openxmlformats.org/package/2006/metadata/core-properties"
    xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:dcterms="http://purl.org/dc/terms/"
    xmlns:dcmitype="http://purl.org/dc/dcmitype/"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <dc:title></dc:title>
    <dc:subject></dc:subject>
    <dc:creator>BlockNote</dc:creator>
    <cp:keywords></cp:keywords>
    <dc:description></dc:description>
    <cp:lastModifiedBy>BlockNote</cp:lastModifiedBy>
    <cp:revision>3</cp:revision>
    <dcterms:created xsi:type="dcterms:W3CDTF">2024-10-19T14:19:00Z</dcterms:created>
    <dcterms:modified xsi:type="dcterms:W3CDTF">2024-11-01T04:50:00Z</dcterms:modified>
</cp:coreProperties>