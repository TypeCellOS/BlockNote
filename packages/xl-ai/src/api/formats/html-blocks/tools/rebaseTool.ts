import { BlockNoteEditor, getBlock } from "@blocknote/core";
import { updateToReplaceSteps } from "../../../../prosemirror/changeset.js";
import {
  getApplySuggestionsTr,
  rebaseTool,
} from "../../../../prosemirror/rebaseTool.js";

export async function createHTMLRebaseTool(
  id: string,
  editor: BlockNoteEditor<any, any, any>,
) {
  const tr = getApplySuggestionsTr(editor);
  const block = getBlock(tr.doc, id);
  if (!block) {
    throw new Error("block not found");
  }

  const html = await editor.blocksToHTMLLossy([
    {
      ...block,
      children: [],
    },
  ]);

  const initialMockID = (window as any).__TEST_OPTIONS?.mockID;

  const blocks = await editor.tryParseHTMLToBlocks(html);

  // hacky
  if ((window as any).__TEST_OPTIONS) {
    (window as Window & { __TEST_OPTIONS?: any }).__TEST_OPTIONS.mockID =
      initialMockID;
  }

  if (blocks.length !== 1) {
    throw new Error("html diff invalid block count");
  }

  const htmlBlock = blocks[0];
  htmlBlock.id = id;

  const steps = updateToReplaceSteps(
    {
      id,
      block: htmlBlock,
    },
    tr.doc,
  );

  if (steps.length) {
    // console.error("html diff", steps);
    throw new Error("html diff");
  }

  return rebaseTool(editor, tr);
}
