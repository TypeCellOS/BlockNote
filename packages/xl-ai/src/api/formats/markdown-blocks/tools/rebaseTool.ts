import { BlockNoteEditor, getBlock } from "@blocknote/core";
import { Mapping } from "prosemirror-transform";
import { updateToReplaceSteps } from "../../../../prosemirror/changeset.js";
import {
  getApplySuggestionsTr,
  rebaseTool,
} from "../../../../prosemirror/rebaseTool.js";

export async function createMDRebaseTool(
  id: string,
  editor: BlockNoteEditor<any, any, any>,
) {
  const tr = getApplySuggestionsTr(editor);
  const md = await editor.blocksToMarkdownLossy([getBlock(tr.doc, id)!]);
  const blocks = await editor.tryParseMarkdownToBlocks(md);

  const steps = updateToReplaceSteps(
    {
      id,
      block: blocks[0],
    },
    tr.doc,
  );

  const stepMapping = new Mapping();
  for (const step of steps) {
    const mapped = step.map(stepMapping);
    if (!mapped) {
      throw new Error("Failed to map step");
    }
    tr.step(mapped);
    stepMapping.appendMap(mapped.getMap());
  }
  return rebaseTool(editor, tr);
}
