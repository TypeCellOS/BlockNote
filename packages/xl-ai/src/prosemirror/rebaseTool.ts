import { BlockNoteEditor } from "@blocknote/core";
import { applySuggestions } from "@blocknote/prosemirror-suggest-changes";
import { Transaction } from "prosemirror-state";
import { Transform } from "prosemirror-transform";

// Helper function to get the transaction that removes all suggestions from the editor.
//
// (rebaseTool is usually used with a transaction that removes all suggestions from the editor)
export function getApplySuggestionsTr(editor: BlockNoteEditor<any, any, any>) {
  let applySuggestionsTr: Transaction;
  applySuggestions(editor.prosemirrorState, (tr) => {
    applySuggestionsTr = tr;
  });

  // @ts-ignore
  if (!applySuggestionsTr) {
    throw new Error("applySuggestionsTr is not set");
  }

  return applySuggestionsTr;
}

/**
 * A helper function to enable rebasing prosemirror operations.
 *
 * The idea is as follows:
 * - There's an update U we want to apply to the editor state E
 * - However, the update U cannot be applied directly to E. For example, because E contains suggestions, or other
 *    formatting incompatible with U
 * - We can however apply U to a "clean" document E', the projection (one that has no suggestions or other formatting)
 * - We can then use the inverse mapping of E' to E to get U' (the rebased update) which can be applied to E
 *
 * This function takes a `projectionTr`, which is the transaction that turns E into E'
 * It returns a helper object that can be used to rebase operations done to E' onto E
 */
export function rebaseTool(
  editor: BlockNoteEditor<any, any, any>,
  projectionTr: Transaction,
) {
  const invertMap = projectionTr.mapping.invert();
  return {
    doc: projectionTr.doc,

    /**
     * Return a new transform that has the projection applied.
     * You can add new operations to this transform and later rebase those on the original document with `rebaseTr`
     */
    tr: () => {
      return new Transform(projectionTr.doc);
    },

    /**
     * Invert map created by the projection.
     * You can use this to map positions on the "clean" document (the projection) to positions on the original document
     */
    invertMap,

    /**
     * Return a new transaction based on the editor state with all the steps you made to `tr`.
     * These steps are rebased on the original document.
     */
    rebaseTr: (tr: Transform) => {
      if (tr.steps.length === 0) {
        throw new Error("No steps to apply");
      }
      let rebasedTr = editor.prosemirrorState.tr;
      for (const step of tr.steps) {
        const mappedStep = step.map(invertMap);
        if (!mappedStep) {
          throw new Error("Step is not mapped");
        }
        rebasedTr = rebasedTr.step(mappedStep);
      }
      return rebasedTr;
    },
  };
}

export type RebaseTool = ReturnType<typeof rebaseTool>;
