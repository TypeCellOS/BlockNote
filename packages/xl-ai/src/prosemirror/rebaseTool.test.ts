import { BlockNoteEditor, getBlockInfo, getNodeById } from "@blocknote/core";
import { expect, it } from "vitest";
import { getApplySuggestionsTr, rebaseTool } from "./rebaseTool.js";

function getExampleEditorWithSuggestions() {
  const editor = BlockNoteEditor.create({
    initialContent: [
      {
        id: "1",
        type: "paragraph",
        content: "HelloHi, world!", // "Hello" will be marked as deleted and "Hi" will be marked as inserted
      },
    ],
  });

  const blockPos = getNodeById("1", editor.prosemirrorState.doc)!;

  const block = getBlockInfo(blockPos);
  if (!block.isBlockContainer) {
    throw new Error("Block is not a container");
  }

  editor.transact((tr) => {
    tr.addMark(
      block.blockContent.beforePos + 1,
      block.blockContent.beforePos + 6,
      editor.pmSchema.mark("deletion", {}),
    );

    tr.addMark(
      block.blockContent.beforePos + 6,
      block.blockContent.beforePos + 8,
      editor.pmSchema.mark("insertion", {}),
    );
  });

  return editor;
}

it("should create some example suggestions", async () => {
  const editor = getExampleEditorWithSuggestions();
  expect(editor.prosemirrorState.doc.toJSON()).toMatchSnapshot();
});

it("should be able to apply changes to a clean doc (use invertMap)", async () => {
  const editor = getExampleEditorWithSuggestions();

  const cleaned = rebaseTool(editor, getApplySuggestionsTr(editor));

  const blockPos = getNodeById("1", cleaned.doc)!;

  const block = getBlockInfo(blockPos);

  if (!block.isBlockContainer) {
    throw new Error("Block is not a container");
  }

  const start = block.blockContent.beforePos + 1;
  const end = start + 2;

  expect(cleaned.doc.textBetween(start, end)).toBe("Hi");

  editor.transact((tr) => {
    tr.replaceWith(
      cleaned.invertMap.map(start),
      cleaned.invertMap.map(end),
      editor.pmSchema.text("What's up"),
    );
  });

  expect(editor.prosemirrorState.doc.toJSON()).toMatchSnapshot();
});

it("should be able to apply changes to a clean doc (use rebaseTr)", async () => {
  const editor = getExampleEditorWithSuggestions();

  const cleaned = rebaseTool(editor, getApplySuggestionsTr(editor));

  const blockPos = getNodeById("1", cleaned.doc)!;

  const block = getBlockInfo(blockPos);

  if (!block.isBlockContainer) {
    throw new Error("Block is not a container");
  }

  const start = block.blockContent.beforePos + 1;
  const end = start + 2;

  expect(cleaned.doc.textBetween(start, end)).toBe("Hi");

  const tr = cleaned.tr();
  tr.replaceWith(start, end, editor.pmSchema.text("What's up"));

  expect(cleaned.rebaseTr(tr).doc.toJSON()).toMatchSnapshot();
});
