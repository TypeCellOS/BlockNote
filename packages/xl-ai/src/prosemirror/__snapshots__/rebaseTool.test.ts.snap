// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to apply changes to a clean doc (use invertMap) 1`] = `
{
  "content": [
    {
      "content": [
        {
          "attrs": {
            "backgroundColor": "default",
            "id": "1",
            "textColor": "default",
          },
          "content": [
            {
              "attrs": {
                "textAlignment": "left",
              },
              "content": [
                {
                  "marks": [
                    {
                      "attrs": {
                        "id": null,
                      },
                      "type": "deletion",
                    },
                  ],
                  "text": "Hello",
                  "type": "text",
                },
                {
                  "text": "What's up, world!",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "blockContainer",
        },
      ],
      "type": "blockGroup",
    },
  ],
  "type": "doc",
}
`;

exports[`should be able to apply changes to a clean doc (use rebaseTr) 1`] = `
{
  "content": [
    {
      "content": [
        {
          "attrs": {
            "backgroundColor": "default",
            "id": "1",
            "textColor": "default",
          },
          "content": [
            {
              "attrs": {
                "textAlignment": "left",
              },
              "content": [
                {
                  "marks": [
                    {
                      "attrs": {
                        "id": null,
                      },
                      "type": "deletion",
                    },
                  ],
                  "text": "Hello",
                  "type": "text",
                },
                {
                  "text": "What's up, world!",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "blockContainer",
        },
      ],
      "type": "blockGroup",
    },
  ],
  "type": "doc",
}
`;

exports[`should create some example suggestions 1`] = `
{
  "content": [
    {
      "content": [
        {
          "attrs": {
            "backgroundColor": "default",
            "id": "1",
            "textColor": "default",
          },
          "content": [
            {
              "attrs": {
                "textAlignment": "left",
              },
              "content": [
                {
                  "marks": [
                    {
                      "attrs": {
                        "id": null,
                      },
                      "type": "deletion",
                    },
                  ],
                  "text": "Hello",
                  "type": "text",
                },
                {
                  "marks": [
                    {
                      "attrs": {
                        "id": null,
                      },
                      "type": "insertion",
                    },
                  ],
                  "text": "Hi",
                  "type": "text",
                },
                {
                  "text": ", world!",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "blockContainer",
        },
      ],
      "type": "blockGroup",
    },
  ],
  "type": "doc",
}
`;
